e2etests/test/twin_disconnect.js

```diff
@@ -95,175 +94,235 @@ var protocolAndTermination = [
+      var deviceClientIdentifier = 'deviceClient';
+      var serviceTwinUpdateIdentifier = 'serviceTwin';
+      rdv.imIn(deviceClientIdentifier);
+      rdv.imIn(serviceTwinUpdateIdentifier);
+
+      var faultInjectionTestPropertyValue = uuid.v4();
       deviceClient.setRetryPolicy(new NoRetry());
-      debug('about to connect a disconnect listener.');
+
+      debug('attaching disconnect event handler');
       deviceClient.on('disconnect', function () {
-        debug('We did get a disconnect message');
-        if (deviceTwin.properties.desired.$version === 2) {
-          testCallback();
+        debug('We did get a disconnect event');
+        if (deviceTwin.properties.desired.testProp === faultInjectionTestPropertyValue) {
+          rdv.imDone(deviceClientIdentifier);
         } else {
-          testCallback(new Error('unexpected disconnect'));
+          testCallback(new Error('unexpected disconnect for device: ' + deviceDescription.deviceId));
         }
       });
-      assert.equal(deviceTwin.properties.desired.$version, 1);
+      debug('attaching desired properties change handler');
+      debug('will trigger fault injection when test property value is: ' + faultInjectionTestPropertyValue);
       deviceTwin.on('properties.desired', function() {
-        if (deviceTwin.properties.desired.$version === 1) {
-          debug('received notification for desired property v1. nothing to do');
-        } else if (deviceTwin.properties.desired.$version === 2) {
+        if (deviceTwin.properties.desired.testProp === faultInjectionTestPropertyValue) {
           var terminateMessage = new Message(' ');
           terminateMessage.properties.add('AzIoTHub_FaultOperationType', testConfiguration.operationType);
           terminateMessage.properties.add('AzIoTHub_FaultOperationCloseReason', testConfiguration.closeReason);
           terminateMessage.properties.add('AzIoTHub_FaultOperationDelayInSecs', testConfiguration.delayInSeconds);
           debug('sending fault injection message');
           deviceClient.sendEvent(terminateMessage, function (sendErr) {
-            debug('at the callback for the fault injection send, err is:' + sendErr);
+            if (sendErr) {
+              debug('fault injection error: ' + sendErr.toString());
+            } else {
+              debug('fault injection successful');
+            }
           });
-        } else if (deviceTwin.properties.desired.$version >= 2) {
-          testCallback(new Error('incorrect property version received - ' + deviceTwin.properties.desired.$version));
+        } else {
+          debug('received notification for desired property change. testProperty: ' + deviceTwin.properties.desired.testProp);
         }
       });
 
       // giving a few seconds for the twin subscription to happen before we send the update.
-      setTimeout(function () {
+      debug('waiting 3 seconds before triggering a twin update from the service API');
```

Comment:
>debug('waiting 3 seconds before triggering a twin update from the service API'); [](start = 6, length = 80)

I get twitchy about this, but I sure don't see a way around it.  Thank you.

Reply:
I'm wondering if we could use the 1st desired properties update as a signal. i'll have to test this out. another day though.